src/components/App/App.js

```diff
@@ -0,0 +1,596 @@
+        ? "Incorrect Password"
+        : "Login Failed. Please Try Again"
+    );
+    console.error(error);
+  };
+
+  //LOGIC FOR HANDLER USER LOGIN
+  const handleLogInModal = () => {
+    setActiveModal("login-signin");
+  };
+  const handleLogInSubmit = (data) => {
+    setIsLoading(true);
+    return auth.logIn(data).then((res) => {
+      if (res.token) {
+        localStorage.setItem("jwt", res.token);
+        setIsLoggedIn(true);
+        auth
+          .checkToken(res.token)
+          .then((user) => {
+            setCurrentUser(user.data);
+            setOracleReadings(user.data.readings);
+            history.push("/");
+            handleCloseModal();
+          })
+          .catch((error) => {
+            setIsLoading(false);
+            if (error.response) {
+              if (error.response.status === 401) {
+                showTinyPopup("Unauthorized: Incorrect credentials");
+              } else if (error.response.status === 500) {
+                showTinyPopup("Server error, try again later");
+              } else {
+                showTinyPopup("An error occurred, please try again");
+              }
+            } else {
+              showTinyPopup(
+                "Network error, please check your connection and try again"
+              );
+            }
+          });
+      }
+    });
+  };
+
+  //LOGIC FOR HANDLING USER LOGOUT
+  const handleLogOut = () => {
+    localStorage.removeItem("jwt");
+    setIsLoggedIn(false);
+    setCurrentUser({});
+    setOracleReadings([]);
+    history.push("/");
+  };
+
+  //LOGIC FOR HANDLING USER REGISTRATION
+  const handleRegisterOnWelcomeModal = () => {
+    if (!isMicActivated) {
+      setIsMicActivationPopupVisible(true);
+    } else {
+      setActiveModal("register-signup");
+    }
+  };
+
+  const handleRegisterModal = () => {
+    setActiveModal("register-signup");
+  };
+
+  const handleRegisterSubmit = (data) => {
+    setIsLoading(true);
+    return auth
+      .register(data)
+      .then((res) => {
+        handleLogInSubmit(data);
+        history.push("/");
+        handleCloseModal();
+      })
+      .catch((error) => {
+        setIsLoading(false);
+        if (error.response.status === 409) {
+          showTinyPopup("User is already registered, please sign in");
+        } else if (error.response.status === 500) {
+          showTinyPopup("Registration failed, please try again");
+        } else {
+          showTinyPopup("error occured on our end, please try again later");
+        }
+      });
+  };
+
+  const handleEditProfileModal = () => {
+    setActiveModal("edit-profile");
+  };
+
+  //LOGIC FOR HANDLING USER EDIT PROFILE
+  const handleEditProfileSubmit = (data) => {
+    setIsLoading(true);
+    return auth
+      .editProfile(data)
+      .then((update) => {
+        setCurrentUser(update.data);
+        handleCloseModal();
+      })
+      .catch(console.error)
+      .finally(() => setIsLoading(false));
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(handleCloseModal) 
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js

  // here is an example
  const handleAddItem = (item) => {
    // here we create a function that returns a promise
    const makeRequest = () => {
      // `return` lets us use a promise chain `then, catch, finally`
      return addNewItem(item).then((item) => {
        setClothingItems([item, ...clothingItems]);
      });
    };
    // here we call handleSubmit passing the request
    handleSubmit(makeRequest);
  };

```

And for the 14th sprint:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```

So, this way you can remove a lot of code duplicating